File: python/mirage/commands/builtin/mongodb/grep.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from functools import partial

from mirage.accessor.mongodb import MongoDBAccessor
from mirage.commands.builtin.generic_bind.search import run_search
from mirage.commands.builtin.mongodb.io import IO
from mirage.commands.config import CommandOpts
from mirage.commands.registry import command
from mirage.commands.spec import SPECS
from mirage.io.types import ByteSource, IOResult
from mirage.types import PathSpec

# The $regex push-down prints each matching document as a whole line;
# pushdown_operand defers shaping flags and multi-operand lines to the
# generic scan, which streams documents rather than reading whole
# collections.
_search = partial(run_search, IO, "grep")


@command("grep", vfs="mongodb", spec=SPECS["grep"])
async def grep(accessor: MongoDBAccessor, paths: list[PathSpec],
               texts: list[str],
               opts: CommandOpts) -> tuple[ByteSource | None, IOResult]:
    return await _search(accessor, paths, texts, opts)

File: typescript/packages/core/src/commands/builtin/langfuse/grep.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { LangfuseAccessor } from '../../../accessor/langfuse.ts'

import { VFSName } from '../../../types.ts'
import { command } from '../../config.ts'
import { specOf } from '../../spec/builtins.ts'
import { runSearch } from '../generic_bind/search.ts'

import { fileReadProvision } from './_provision.ts'
import { LANGFUSE_IO } from './io.ts'

export const LANGFUSE_GREP = command({
  name: 'grep',
  vfs: VFSName.LANGFUSE,
  spec: specOf('grep'),
  fn: (accessor: LangfuseAccessor, paths, texts, opts) =>
    runSearch(LANGFUSE_IO, 'grep', accessor, paths, texts, opts),
  provision: fileReadProvision,
})
